File: js/src/connection.ts
/** GQL connection management. */

import { Client, credentials, type ChannelCredentials } from "@grpc/grpc-js";
import { SessionServiceClient, GqlServiceClient } from "./generated/gql_service";
import { DatabaseClient } from "./database";
import { GqlSession } from "./session";

/** A connection to a GWP server. */
export class GqlConnection {
  private readonly sessionClient: SessionServiceClient;
  private readonly gqlClient: GqlServiceClient;
  private readonly _endpoint: string;
  private readonly _creds: ChannelCredentials;

  private constructor(
    endpoint: string,
    creds: ChannelCredentials,
    sessionClient: SessionServiceClient,
    gqlClient: GqlServiceClient,
  ) {
    this._endpoint = endpoint;
    this._creds = creds;
    this.sessionClient = sessionClient;
    this.gqlClient = gqlClient;
  }

  /**
   * Connect to a GWP server.
   * @param endpoint Server address (e.g. "localhost:50051").
   * @param credentials Optional channel credentials. Defaults to insecure.
   */
  static connect(
    endpoint: string,
    channelCredentials?: ChannelCredentials,
  ): GqlConnection {
    const creds = channelCredentials ?? credentials.createInsecure();
    const sessionClient = new SessionServiceClient(endpoint, creds);
    const gqlClient = new GqlServiceClient(endpoint, creds);
    return new GqlConnection(endpoint, creds, sessionClient, gqlClient);
  }

  /** Perform handshake and return a session. */
  async createSession(): Promise<GqlSession> {
    return GqlSession.create(this.sessionClient, this.gqlClient);
  }

  /** Create a database management client. */
  createDatabaseClient(): DatabaseClient {
    return new DatabaseClient(this._endpoint, this._creds);
  }

  /** Close the underlying gRPC channels. */
  close(): void {
    Client.prototype.close.call(this.sessionClient);
    Client.prototype.close.call(this.gqlClient);
  }
}

File: go/connection.go
package gwp

import (
	"context"

	"google.golang.org/grpc"
	"google.golang.org/grpc/credentials/insecure"

	pb "github.com/GrafeoDB/gql-wire-protocol/go/gen/gql"
)

// GqlConnection is a connection to a GWP server.
type GqlConnection struct {
	conn          *grpc.ClientConn
	sessionClient pb.SessionServiceClient
	gqlClient     pb.GqlServiceClient
}

// Connect creates a new connection to a GWP server.
func Connect(ctx context.Context, target string, opts ...grpc.DialOption) (*GqlConnection, error) {
	if len(opts) == 0 {
		opts = append(opts, grpc.WithTransportCredentials(insecure.NewCredentials()))
	}

	conn, err := grpc.NewClient(target, opts...)
	if err != nil {
		return nil, &GqlError{Message: "failed to connect: " + err.Error()}
	}

	return &GqlConnection{
		conn:          conn,
		sessionClient: pb.NewSessionServiceClient(conn),
		gqlClient:     pb.NewGqlServiceClient(conn),
	}, nil
}

// CreateSession performs a handshake and returns a new session.
func (c *GqlConnection) CreateSession(ctx context.Context) (*GqlSession, error) {
	resp, err := c.sessionClient.Handshake(ctx, &pb.HandshakeRequest{
		ProtocolVersion: 1,
	})
	if err != nil {
		return nil, err
	}

	if resp.SessionId == "" {
		return nil, &SessionError{Message: "server returned empty session ID"}
	}

	return &GqlSession{
		sessionID:     resp.SessionId,
		sessionClient: c.sessionClient,
		gqlClient:     c.gqlClient,
	}, nil
}

// Close closes the underlying gRPC connection.
func (c *GqlConnection) Close() error {
	return c.conn.Close()
}
